### PR TITLE
Optional reshape of predictions in Perplexity metric

### PR DESCRIPTION
**Pull request by fhieber**
Fixes a shape issue with the Perplexity metric after #7949  and #8003 were merged: after these PRs, the executor group slices graph outputs according to the size of the batch axis. This forces graph/symbol outputs to be batch-major if used to compute metrics.
In cases like Sockeye, we used to feed label as `(batch_size, length)` into the graph, but return `(batch_size*length, vocab_size)` as output after SoftmaxOutput. This is also the required format for the Perplexity metric.
With the PRs mentioned above, this now fails, as the executor_group slices the `(batch_size*length, vocab_size)` output to `(batch_size, vocab_size)`, which is obviously wrong.

We can fix this by using `preserve_shape=True` in our SoftmaxOutput call and return `(batch_size, length, vocab_size)` outputs.
However, this caused the Perplexity metric to fail.

This PR fixes this by doing an optional reshape of the predictions to 2d.

### CODE REVIEW

**Comment by piiswrong** [comment]
I have reverted that PR but I think this PR is still needed

File: python/mxnet/metric.py [hunk 1]
```diff
@@ -646,9 +646,13 @@ def update(self, labels, preds):
...
         num = 0
         for label, pred in zip(labels, preds):
-            assert label.size == pred.size/pred.shape[-1], \
+            k = pred.shape[self.axis]
+            n = int(pred.size/k)
+            assert label.size == n, \
                 "shape mismatch: %s vs. %s"%(label.shape, pred.shape)
             label = label.as_in_context(pred.context).reshape((label.size,))
+            if len(pred.shape) > 2:
+                pred = pred.reshape((n, k))
```
Comment (piiswrong) [inline python/mxnet/metric.py]: This wouldn't work if axis is not -1

Comment (fhieber) [reply python/mxnet/metric.py]: I also realized this when making the change. I will try to make some time today to look into it, but I will be OOTO for the next 10 days after.

**Comment by piiswrong** [comment]
This is fine but in case you care about performance pick actually supports picking on axes in the middle.

It works if you have pred.shape = (x, y, z) and label.shape = (x, 1, z) and axis= 1

**Comment by piiswrong** [comment]
@reminisce Could you take over? I think @zackchase complained about metric speed

**Comment by eric-haibin-lin** [comment]
What's the next step for this PR? @reminisce @piiswrong

**Comment by reminisce** [comment]
Sorry I didn't notice that I was tagged.
@piiswrong Did you mean changing this line into using `pick`? I remember @szha was working on improving the metric speed.

**Comment by szha** [comment]
Yes, I was looking into the metrics, and I didn't feel like having a separate perplexity metric was necessary in gluon. The reason is that perplexity is usually just the exp(cross_entropy_loss). Since cross_entropy_loss is required for training and often used in testing, getting perplexity is just about calculating its exponential. Having a separate perplexity in that case could mean unnecessarily repeated computation.

**Comment by CodingCat** [comment]
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!
